neat_core.c

```diff
@@ -486,14 +649,14 @@ static int io_readable(neat_ctx *ctx, neat_flow *flow,
     struct sctp_recvv_rn rn;
     socklen_t infolen = sizeof(struct sctp_recvv_rn);
     int flags = 0;
-#endif
-#endif
+#endif // else !defined(USRSCTP_SUPPORT)
+#endif // defined(SCTP_SUPPORTED)
     neat_log(NEAT_LOG_DEBUG, "%s", __func__);
 
     if (!flow->operations || !flow->operations->on_readable) {
         return READ_WITH_ERROR;
     }
-#if defined(IPPROTO_SCTP)
+#if defined(SCTP_SUPPORTED)
     if ((flow->sockProtocol == IPPROTO_SCTP) &&
```

Comment:
> On 03 Jun 2016, at 14:47, Kristian A. Hiorth <notifications@github.com> wrote:
> 
> In neat_core.c:
> 
> >      neat_log(NEAT_LOG_DEBUG, "%s", __func__);
> >  
> >      if (!flow->operations || !flow->operations->on_readable) {
> >          return READ_WITH_ERROR;
> >      }
> > -#if defined(IPPROTO_SCTP)
> > +#if defined(SCTP_SUPPORTED)
> >      if ((flow->sockProtocol == IPPROTO_SCTP) &&
> 
> Ok, that does make sense for this particular code path. In the bigger picture, I'm a bit worried that things will still be broken elsewhere because of assumptions around what to do when SCTP support is detected or not.
> 
> Is there any consensus on how this should work? I see for example that on OSX, HE thinks it can try SCTP, although that isn't really supported. That's different from NetBSD, where it only detects TCP.
And that is perfectly fine. HE is for detecting what is working and what not.
> 
> Also, what happens if several different SCTP implementations are available?
Currently you can't compile the userland stack in when a kernel stack is also available.
A fix is on the ToDo list, but has low priority. On FreeBSD I would always prefer
the kernel stack. On Linux it might depend. But we have to deal with this in the
future. Right now our assumption was that the userland stack is only compiled in
when no kernel stack is available.
> Right now, all code paths will always prefer USRSCTP if that is compiled in if I understand the code correctly.
Yes. Right now you can only have one. It is a limitation.

Please note that you can't do native SCTP in userland if the kernel supports SCTP. So the
only option for having both is on Linux and
* ensure that the SCTP module is not loaded (I think it is automatically loaded when opening
  an SCTP socket) and use the userland stack for SCTP/IP and SCTP/UDP/IP.
* use the kernel stack for SCTP/IP and the userland stack for SCTP/UDP/IP.

Best regards
Michael
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>